panther_hardware_interfaces/src/canopen_controller.cpp

```diff
@@ -0,0 +1,181 @@
+{
+  canopen_communication_started_.store(false);
+
+  canopen_communication_thread_ = std::thread([this]() {
+    ConfigureRT();
+
+    try {
+      InitializeCanCommunication();
+    } catch (const std::system_error & e) {
+      std::cerr << "Exception caught during CAN initialization: " << e.what() << std::endl;
+
+      NotifyCanCommunicationStarted(false);
+      return;
+    }
+
+    NotifyCanCommunicationStarted(true);
+
+    try {
+      loop_->run();
+    } catch (const std::system_error & e) {
+      // If the error happens and loop stops SDO and PDO operations will timeout and in result
+      // system will switch to error state
+      std::cerr << "Exception caught in loop run: " << e.what() << std::endl;
+    }
+  });
+
+  if (!canopen_communication_started_.load()) {
+    std::unique_lock<std::mutex> lck(canopen_communication_started_mtx_);
+    canopen_communication_started_cond_.wait(lck);
+  }
+
+  if (!canopen_communication_started_.load()) {
+    throw std::runtime_error("CAN communication not initialized");
+  }
+
+  BootDrivers();
+}
+
+void CanOpenController::Deinitialize()
+{
+  if (master_) {
+    master_->AsyncDeconfig().submit(*exec_, [this]() { ctx_->shutdown(); });
+  }
+  canopen_communication_thread_.join();
+  canopen_communication_started_.store(false);
+
+  rear_driver_.reset();
+  front_driver_.reset();
+  master_.reset();
+  chan_.reset();
+  ctrl_.reset();
+  timer_.reset();
+  exec_.reset();
+  loop_.reset();
+  poll_.reset();
+  ctx_.reset();
+}
+
+void CanOpenController::InitializeCanCommunication()
+{
+  lely::io::IoGuard io_guard;
+
+  ctx_ = std::make_shared<lely::io::Context>();
+  poll_ = std::make_shared<lely::io::Poll>(*ctx_);
+  loop_ = std::make_shared<lely::ev::Loop>(poll_->get_poll());
+  exec_ = std::make_shared<lely::ev::Executor>(loop_->get_executor());
+
+  timer_ = std::make_shared<lely::io::Timer>(*poll_, *exec_, CLOCK_MONOTONIC);
+
+  ctrl_ = std::make_shared<lely::io::CanController>("panther_can");
+  chan_ = std::make_shared<lely::io::CanChannel>(*poll_, *exec_);
+
+  chan_->open(*ctrl_);
+
+  // Master dcf is generated from roboteq_motor_controllers_v80_21 using following command:
+  // dcfgen panther_can.yaml -r
+  // dcfgen comes with lely, -r option tells to enable remote PDO mapping
+  std::string master_dcf_path = std::filesystem::path(ament_index_cpp::get_package_share_directory(
+                                  "panther_hardware_interfaces")) /
+                                "config" / "master.dcf";
+
+  master_ = std::make_shared<lely::canopen::AsyncMaster>(
+    *timer_, *chan_, master_dcf_path, "", canopen_settings_.master_can_id);
+
+  front_driver_ = std::make_shared<RoboteqDriver>(
+    *exec_, *master_, canopen_settings_.front_driver_can_id,
+    canopen_settings_.sdo_operation_timeout);
+  rear_driver_ = std::make_shared<RoboteqDriver>(
+    *exec_, *master_, canopen_settings_.rear_driver_can_id,
+    canopen_settings_.sdo_operation_timeout);
```

Comment:
```suggestion
  front_driver_ = std::make_shared<RoboteqDriver>(
    exec_, master_, canopen_settings_.front_driver_can_id,
    canopen_settings_.sdo_operation_timeout);
  rear_driver_ = std::make_shared<RoboteqDriver>(
    exec_, master_, canopen_settings_.rear_driver_can_id,
    canopen_settings_.sdo_operation_timeout);
```
This suggestion is reflecting possible changes in `RoboteqDriver` class